iot/api-client/end_to_end_example/cloudiot_pubsub_example_server.py

```diff
@@ -0,0 +1,211 @@
+    ).modifyCloudToDeviceConfig(
+        name=device_name, body=body)
+
+    # The http call for the device config change is thread-locked so that we
+    # don't have competing threads simultaneously using the httplib2 library,
+    # which is not thread-safe.
+    self._update_config_mutex.acquire()
+    try:
+      request.execute()
+    except HttpError, e:
+      # If the server responds with a HtppError, we log it here, but continue
+      # so that the message does not stay NACK'ed on the pubsub channel.
+      print 'Error executing ModifyCloudToDeviceConfig: {}'.format(e)
+    finally:
+      self._update_config_mutex.release()
+
+  def run(self, project_id, pubsub_subscription):
+    """The main loop. Consumes messages from the Pub/Sub subscription."""
+
+    subscriber = pubsub_v1.SubscriberClient()
+    subscription_path = subscriber.subscription_path(project_id,
+                                                     pubsub_subscription)
+
+    def callback(message):
+      """Logic executed when a message is received from subscribed topic."""
+      try:
+        data = json.loads(message.data)
+      except ValueError, e:
+        print 'Loading Payload ({}) threw an Exception: {}.'.format(
+            message.data, e)
+        message.ack()
+        return
+
+      # Get the registry id and device id from the attributes. These are
+      # automatically supplied by IoT, and allow the server to determine which
+      # device sent the event.
+      device_project_id = message.attributes['projectId']
+      device_registry_id = message.attributes['deviceRegistryId']
+      device_id = message.attributes['deviceId']
+      device_region = message.attributes['deviceRegistryLocation']
+
+      # Send the config to the device.
+      self._update_device_config(device_project_id, device_region,
+                                 device_registry_id, device_id, data)
+
+      # Acknowledge the consumed message. This will ensure that they are not
+      # redelivered to this subscription.
+      message.ack()
+
+    print 'Listening for messages on {}'.format(subscription_path)
+    subscriber.subscribe(subscription_path, callback=callback)
+
+    # The subscriber is non-blocking, so we must keep the main thread from
+    # exiting to allow it to process messages in the background.
+    while True:
+      time.sleep(60)
+
+
+def parse_command_line_args():
+  """Parse command line arguments."""
+  parser = argparse.ArgumentParser(
+      description='Example of Google Cloud IoT registry and device management.')
+  # Required arguments
+  parser.add_argument(
+      '--project_id', required=True, help='GCP cloud project name.')
+  parser.add_argument(
+      '--pubsub_subscription',
+      required=True,
+      help='Google Cloud Pub/Sub subscription name.')
+  parser.add_argument('--api_key', required=True, help='Your API key.')
+
+  # Optional arguments
+  parser.add_argument(
```

Comment:
We generally default to the GOOGLE_APPLICATION_CREDENTIALS environment variable for this

Reply:
Done.